metronome/checkpointing/service/src/io/iohk/metronome/checkpointing/service/CheckpointingService.scala

```diff
@@ -30,19 +32,38 @@ import scala.annotation.tailrec
 
 class CheckpointingService[F[_]: Sync, N](
     ledgerTreeRef: Ref[F, LedgerTree],
+    mempoolRef: Ref[F, Mempool],
     lastCommittedHeaderRef: Ref[F, Block.Header],
     checkpointDataRef: Ref[F, Option[CheckpointData]],
-    //TODO: PM-3137, this is used for testing that a certificate was created correctly
-    //      replace with proper means of pushing the certificate to the Interpreter
-    pushCheckpointFn: CheckpointCertificate => F[Unit],
     ledgerStorage: LedgerStorage[N],
-    blockStorage: BlockStorage[N, CheckpointingAgreement]
+    blockStorage: BlockStorage[N, CheckpointingAgreement],
+    interpreterClient: InterpreterRPC[F]
 )(implicit storeRunner: KVStoreRunner[F, N])
     extends ApplicationService[F, CheckpointingAgreement] {
 
   override def createBlock(
       highQC: QuorumCertificate[CheckpointingAgreement]
-  ): F[Option[Block]] = ???
+  ): F[Option[Block]] = (
+    for {
+      parent    <- OptionT(getBlock(highQC.blockHash))
+      oldLedger <- OptionT(projectLedger(parent))
+      mempool   <- OptionT.liftF(projectMempool)
+
+      newBody <- OptionT {
+        if (mempool.isEmpty)
+          Block.Body.empty.some.pure[F]
+        else
+          interpreterClient.createBlockBody(oldLedger, mempool.proposerBlocks)
+      }
+
+      newLedger = oldLedger.update(newBody.transactions)
+      newBlock  = Block.make(parent.header, newLedger.hash, newBody)
+
+      _ <- OptionT.liftF(
+        updateLedgerTree(newLedger, newBlock, precomputed = true)
+      )
```

Comment:
Once the block is created, you can set `hasNewCheckpointCandidate` to false.

Reply:
What if another checkpoint notification comes along while creating the block? It's probably not a big issue since it's a really tiny window where it cause a race condition and if it does other federation members should continue including checkpoint candidates, but I'm not entirely happy relying on that. Should we add locking?

Reply:
Good point! I think the solution is to use [Ref.getAndUpdate](https://github.com/typelevel/cats-effect/blob/c86425cbc88a5d1551ba26a6b76a8cac1d493002/kernel/shared/src/main/scala/cats/effect/kernel/Ref.scala#L47) instead of `Ref.get` to access the mempool, and set the "dirty" flag right there when it's read. This would be in `projectMempool` before any further filtering is done, but then if another notification comes in, it's fine because we don't need to update the mempool again during block creation.

Reply:
Related to this is at which point we should save the filtered `Mempool` in its `Ref`, so that we don't lose proposerBlocks that may come along?

What if we had an effectful `projectMempool(hash: Block.Hash): F[Mempool]` and a pure `filterMempool(body: Block.Body)(mempool: Mempool): Mempool`. `projectMempool` uses `filterMempool` but doesn't save the result. Then for each executed block we do `mempoolRef.update(filterMempool(body))`. WDYT?

Reply:
Also a very good question, something I've been pondering myself over the weeks too. 

I think that removing transactions from the mempool after executing a block is the right thing to do. These transactions either got checkpointed or they ended up in the ledger. We can't do it after creating a block because it may not gather enough votes, and we have to re-propose the same transaction next time. 

There are some potential problems, though, which I mentioned to @fishermanymc. You asked last week how we should deal with proposed transactions that we have already seen somehow, and I said they should be handled by the Interpreter which actually understands the content. For example we could be at height 101, having checkpointed height 100 already (say, every 10th), and someone proposes to include a transaction and height 75. The Interpreter can just reject such block because maybe Advocate only allows uncle blocks that are newer than the last checkpoint, or the one before that (level 100 or 90). 

But what if I'm an honest node and for some reason I just made a block at level 89 and it got into my mempool, how to get rid of it now? Should there be a TTL, in terms of proposal attempts? Or maybe we should have one more message between the Service and the Interpreter, where we actually hand over the mempool to the Interpreter after each committed or executed block to be filtered?